Fixes #14067 - Host#edit integration tests should not wait for tasks or show graphs

dLobatog opened the pull request:
When you click on Host#edit Submit button, a few orchestration tasks
will be created. After that, the show page will call (via AJAX) a few
URLS to fetch 'overview', 'resources', 'nics', etc... to fill up the
table on the left side. These are all irrelevant to tests that only
concern the form, and pose a problem.

**Problem**
The server we test on is not multithreaded, these tasks will run
sequentially. Therefore, the default Capybara waiting time, or
Capybara.default_max_time might not be enough to wait for all requests.

**Solution**
We can stub out these calls, and wait until the #host-progress div has
disappeared for POST/PATCH requests.

---

This should fix some of the spotty tests in http://projects.theforeman.org/issues/13293, specifically 'correctly override global params' - 'class parameters and overrides are displayed correctly for booleans' - some of the other methods "fixed" here also were prone to fail.

Reproduce the failures on host_integration by adding this to ApplicationController, and reducing `Capybara.default_max_wait_time` to 10.

``` ruby
  after_filter -> do
    `sleep 5`
    puts "#{request.method} #{request.url} took #{Time.now - @@req_time} to complete"
  end
  before_filter -> { @@req_time = Time.now }
```

dLobatog commented on test/test_helper.rb:
```diff
@@ -331,6 +331,24 @@ def wait_for_ajax
+    def wait_for_orchestration_requests
+      Timeout.timeout(Capybara.default_max_time * 2) do
+        # Notice host-progress won't be visible without a multi-threaded server
+        loop { find("#host-progress", :visible => false, :wait => 0.1) }
```
I could not find a way to do this w/o rescuing for a not found exception

domcleal commented:
There are further test failures, could you look please?

<pre>
    HostIntegrationTest::edit page.test_0002_can override puppetclass lookup values
    HostIntegrationTest::create new host page.test_0005_sets fields to "inherit" when hostgroup is selected
    HostIntegrationTest::edit page.test_0001_class parameters and overrides are displayed correctly for strings
    HostIntegrationTest::edit page.test_0004_correctly override global params
    HostIntegrationTest::edit page.test_0006_fields are not inherited on edit
    HostIntegrationTest::edit page.test_0008_class parameters and overrides are displayed correctly for booleans
</pre>

dLobatog commented:
@domcleal Are you able to reproduce any of these locally? Stubbing the methods so that they won't render anything works fine here with `rake test:integration`, `jenkins:integration`, or plain `ruby -I'test' test/integration/host_test.rb`

domcleal commented:
Yes, I ran `be ruby -Itest test/integration/host_test.rb -n /inherit/` and reproduced it:

<pre>
  1) Error:
HostIntegrationTest::create new host page#test_0005_sets fields to "inherit" when hostgroup is selected:
ActionView::MissingTemplate: Missing template hosts/nics, application/nics with {:locale=>[:en], :formats=>[:html, :html, :text, :js, :css, :ics, :csv, :vcf, :png, :jpeg, :gif, :bmp, :tiff, :mpeg, :xml, :rss, :atom, :yaml, :multipart_form, :url_encoded_form, :json, :pdf, :zip], :variants=>[], :handlers=>[:erb, :builder, :raw, :ruby, :rabl]}. Searched in:
  * "/home/dcleal/code/foreman/foreman/app/views"
</pre>

I don't think returning `true` from those controller action stubs is enough, since Rails will still use implicit rendering (which doesn't work as we don't have corresponding views).  You would need to actually call `render :nothing => true` or similar to prevent rendering.

Is it possible to actually disable the Javascript from making those AJAX calls rather than trying to partially disable them in the controllers?  Note that we can load extra JS (like test/integration/support/disable_animations.js) so could potentially simplify or disable this functionality.  Stubbing the actions and rendering different empty pages or incomplete responses might also trigger errors in the client-side AJAX, leading to further bugs...

dLobatog commented:
Still can't reproduce it with that, Ruby 2.2.3, tried on pg and sqlite. 

I'm trying out https://github.com/bblimke/webmock now to mock the call much earlier,  as soon as it hits the server. As for stubbing the actual `.ajax` method, I guess I could introduce http://sinonjs.org/ or something similar, but I'm not sure it's the way to go.

domcleal commented:
I'm not sure that webmock would help, it's usually used to stub on the (Ruby) client side, not the server, or in browser AJAX clients.

dLobatog commented:
You're right, it won't do anything. Let's just skip the stubbing and wait for the orchestration requests at least, which would make some tests more solid (the ones that use orchestration)

domcleal commented:
Just going back to the commit message and description (which need updating I think, since it _does_ wait), should wait_for_orchestration calls be followed by a wait_for_ajax?

On submitting the host form it'll trigger the orchestration/tasks progress monitoring, and then load the graphs on the host#show page.  If those calls aren't stubbed they could also take a while, but if wait_for_orchestration is followed by the existing wait_for_ajax call, then it should wait for those to complete too.

domcleal commented on test/test_helper.rb:
```diff
@@ -331,6 +331,15 @@ def wait_for_ajax
+    def wait_for_orchestration_requests
+      Timeout.timeout(Capybara.default_max_wait_time * 2) do
+        # Notice host-progress won't be visible without a multi-threaded server
+        loop { find("#host-progress", :visible => false, :wait => 0.1) }
```
Is this whole loop equivalent to `assert has_no_selector?`?  It just waits until the `#host-progress` element is no longer present.

dLobatog committed:
Fixes #14067 - Host#edit integration tests should wait for tasks

When you click on Host#edit Submit button, a few orchestration tasks
will be created. After that, the show page will call (via AJAX) a few
URLS to fetch 'overview', 'resources', 'nics', etc... to fill up the
table on the left side. These are all irrelevant to tests that only
concern the form, and pose a problem as we don't explicitly wait for
them.

Problem:
The server we test on is not multithreaded, these tasks will run
sequentially. Therefore, the default Capybara waiting time, or
Capybara.default_max_time might not be enough to wait for all requests.

Solution:
We can wait until the #host-progress div has disappeared and wait for
the Host#show ajax calls using wait_for_ajax

dLobatog commented:
I've updated this one to depend on #3293 - it'll change significantly after that so let's leave this one as is until that one is merged

domcleal commented:
The other's merged, could you please rebase?

domcleal commented on test/integration/shared/host_orchestration_stubs.rb:
```diff
@@ -0,0 +1,19 @@
+module HostOrchestrationStubs
+  extend ActiveSupport::Concern
+
+  def stub_show_ajax_requests
```
unused?

domcleal commented on test/integration/host_js_test.rb:
```diff
@@ -167,7 +172,7 @@ class HostJSTest < IntegrationTestWithJavascript
         loop while find(:css, '#interfaceModal', :visible => false).visible?
       end
       click_button 'Submit' #create new host
-      wait_for_ajax
```
Should we keep this in?  It might help wait for some of the graphs on the host#show page.

Edit: and if so, perhaps use it in all places we submit the host form - perhaps add a helper method into the new concern?

domcleal commented:
The commit message also still mentions that graphs are stubbed, which they're not.  I added an inline comment about perhaps keeping wait_for_ajax in at the same time as waiting for progress bars - it might help wait for them?

dLobatog commented:
@domcleal No luck with stubbing yet - I tried with `render nothing` and `render :text => 'foo'` but still implicit render would automatically trigger and raise these MissingTemplate errors we've seen.

To disable AJAX loading from the js side, I tried removing the ajax tags. On development it would work, but on Poltergeist it would not trigger `onreadystatechange` after clicking on Submit (it triggers it after all other requests):

```
document.onreadystatechange = function() {
  if (document.readyState === "interactive") {
    ajax_loading_elements = document.querySelectorAll('div[data-ajax-url]');
    [].forEach.call(ajax_loading_elements, function(div) {
      div.removeAttribute('data-ajax-url');
    })
  }
}
```

I tried listening for other events like DOMContentLoaded without success too, not even `window.onload` stuff seems to load (according to `console.log`) after clicking on Submit. 

Any ideas? I found https://github.com/oesmith/puffing-billy which looks like would work but looks like overkill for this

domcleal commented:
Sorry, to be clear - I'm OK without the stubbing, but perhaps best to remove it from the commit message and continue without it for now.  Would wait_for_ajax at least wait for those requests to complete?

dLobatog commented:
@domcleal `wait_for_ajax` might help, depending on how slow the requests are. What I think we should do if we don't stub these out, is to write a method `click_on_submit`. There we click and wait for ajax and for #host-progress to disappear. Hopefully that'll make it more stable. I've updated the commit message to explain we're just explicitly waiting here, not stubbing anything

domcleal commented on test/integration/shared/host_orchestration_stubs.rb:
```diff
@@ -0,0 +1,19 @@
+      # Notice host-progress won't be visible without a multi-threaded server
+      loop until has_no_selector?("#host-progress", :visible => false, :wait => 0.1)
+    end
+  rescue Capybara::ElementNotFound
```
I ran a few jobs on Jenkins to test this, one failed with this:

<pre>
Capybara::Poltergeist::ObsoleteNode: The element you are trying to interact with is either not part of the DOM, or is not currently visible on the page (perhaps display: none is set). It's possible the element has been replaced by another element and you meant to interact with the new element. If so you need to do a new 'find' in order to get a reference to the new element.
    test/integration/shared/host_orchestration_stubs.rb:14:in `block in wait_for_orchestration_requests'
    test/integration/shared/host_orchestration_stubs.rb:12:in `wait_for_orchestration_requests'
    test/integration/shared/host_orchestration_stubs.rb:6:in `click_on_submit'
    test/integration/host_js_test.rb:168:in `block (2 levels) in <class:HostJSTest>' (Capybara::Poltergeist::ObsoleteNode)
/usr/local/rvm/gems/ruby-2.1.5@test_develop_pr_core-0/gems/capybara-2.6.2/lib/capybara/node/matchers.rb:53    
</pre>


(http://ci.theforeman.org/job/test_develop_pr_core/8566/database=postgresql,label=fast,ruby=2.1/)

dLobatog replied:
Hmm from what I'm reading that happens if the element is set to `display:none`. The default class for #host-progress (hide) is equivalent to that. I'll simplify this a bit though, I don't think the Timeout and rescue are needed.

dLobatog commented:
Hmm.. the failure on http://ci.theforeman.org/job/test_develop_pr_core/8569/database=postgresql,label=fast,ruby=2.0.0/console isn't about submitting a Host, but something on the parameters page. I'm trying to reproduce that failure now, although it's unrelated to the fixes here.

domcleal commented:
Yes, it does seem to be unrelated, but I believe the number of those particular failures has increased a lot since #14097.  I'll try to have a look as well next week.

Merged as 1580adb, thanks @dLobatog!

domcleal commented:
Here's another possible approach to stubbing those AJAX requests, by using poltergeist's URL blacklist: https://github.com/domcleal/foreman/commit/65ea613551982c889099673791602f59f2e844b9

My only concern is that we then don't get to test the AJAX on this page at all (though perhaps we could add a test without the blacklisting) and that the blacklisting is a very basic "URL contains X" check, so the URL fragments aren't very precise.
